plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/thrift/TokenFetchingMetastoreClientFactory.java

```diff
@@ -70,15 +73,25 @@ public ThriftMetastoreClient createMetastoreClientFor(Optional<ConnectorIdentity
         String username = identity.map(ConnectorIdentity::getUser)
                 .orElseThrow(() -> new IllegalStateException("End-user name should exist when metastore impersonation is enabled"));
 
-        String delegationToken;
         try {
-            delegationToken = delegationTokenCache.getUnchecked(username);
+            return clientProvider.createMetastoreClient(Optional.of(getDelegationToken(username)));
+        }
+        catch (TException e) {
```

Comment:
can we be look deeper into exception to determine if regenerating delegatin token can really be helpful?

Reply:
An exception is occurring as below.
#### Trino Error (402 version)
<img width="1270" alt="스크린샷 2022-11-08 오후 12 29 12" src="https://user-images.githubusercontent.com/62536685/200471148-4a8e50d5-27bf-4fe1-ac94-d19dc1ae6236.png">

#### Hive Metastore Error (3.1.2 version)
<img width="1507" alt="스크린샷 2022-11-08 오후 12 32 48" src="https://user-images.githubusercontent.com/62536685/200471294-0f3ef654-9f2d-4f7b-bf17-12a4d0b1637e.png">

Reply:
Maybe sth like that
```java
    @Override
    public ThriftMetastoreClient createMetastoreClientFor(Optional<ConnectorIdentity> identity)
            throws TException
    {
        if (!impersonationEnabled) {
            return createMetastoreClient();
        }

        String username = identity.map(ConnectorIdentity::getUser)
                .orElseThrow(() -> new IllegalStateException("End-user name should exist when metastore impersonation is enabled"));

        try {
            return clientProvider.createMetastoreClient(Optional.of(getDelegationToken(username)));
        }
        catch (TException e) {
            if (isDelegationTokenExpiration(e)) {
                // try once again with new token
                String delegationToken = loadDelegationToken(username);
                delegationTokenCache.put(username, delegationToken);
                return clientProvider.createMetastoreClient(Optional.of(delegationToken));
            }
            throw e;
        }
    }

    private boolean isDelegationTokenExpiration(Throwable exception)
    {
        if (exception instanceof SecretManager.InvalidToken) {
            return true;
        }
        if (exception.getCause() == null || exception.getCause() == exception) {
            return false;
        }
        return (isDelegationTokenExpiration(exception.getCause()));
    }

```

Reply:
Unfortunately, the message passed to the thrift response cannot be identified as `SecretManager.InvalidToken` because it is transmitted only with the following string information.

* [TSaslTransport](https://github.com/apache/thrift/blob/e8353cb46e9f5e71f9b76f55d6bf59530b7f98ef/lib/java/src/main/java/org/apache/thrift/transport/TSaslTransport.java#L164-L167)
<img width="1129" alt="스크린샷 2022-11-09 오전 1 50 43" src="https://user-images.githubusercontent.com/62536685/200630224-c75cd8cc-2a21-4a2a-b9bf-26e4d20173d5.png">

Reply:
Oh - I misread your previous message. The "cause" does not make it through the "TTransportException".  :/

Reply:
Ok - let's go with that. Just please add a comment in `catch` clause why are you reloading the token.

Reply:
> Just please add a comment in catch clause why are you reloading the token.

Added a comment to that.

> Also please add a code which ensures that we are not refreshing the token "too often", I think ensuring at most 1 refresh per minute will be fine. We do not want to start hammering metastore which returns errors, because it is overloaded with twice as many requests.

I also positively agree with that.
Since tokens are generated for each user and each expires, the refresh period check is implemented to be checked for each user.